### ISSUE TITLE
[Bug] Mapped-layer courier can reuse CUDA storage before compute finishes

### ISSUE THREAD

**Issue by Grypse** [sgl-project/sglang#37963]
## Checklist

- [x] I searched related issues and pull requests but found no solution.
- [x] The bug persists in the latest `main` source (`dae126d510b786a2424684c17ca4e5e70d74e8f7`).
- [x] Environment info and a minimal reproducer are included below.
- [x] This is a bug report, not a general question.

## Describe the bug

The diffusion layerwise-offload mapped-weight courier creates CUDA tensors on its private copy stream, then returns them to the compute thread. `_collect_mapped_layer()` waits for the copy event, but it does not call `gpu_tensor.record_stream(compute_stream)`.

After a layer's forward is enqueued, the post-hook replaces the parameter data with a placeholder. When the last Python reference to the courier tensor is released, PyTorch's caching allocator only knows about the courier stream. Because the copy has completed there, it may reuse the storage while a compute-stream kernel is still reading it. This silently corrupts weights and produced colored-noise MiniMax-H3 videos without an OOM, NaN, or task failure.

`wait_event()` provides execution ordering, but it does not register the tensor's lifetime on the consuming stream. The consuming stream must also be recorded.

The same missing lifetime registration is present in current `main`:

```python
with torch.inference_mode(False), torch.no_grad():
    for name, gpu_tensor in tensors.items():
        target = self.get_target_with_name(name)
        target.data = self._wrap_for_target(target, gpu_tensor)
torch.get_device_module().current_stream().wait_event(event)
```

## Reproduction

The full-model failure was reproduced with `MiniMaxAI/MiniMax-H3` FL2VA, layerwise offload, and mapped weights:

```bash
SGLANG_DIFFUSION_DISABLE_MAPPED_COURIER=0 \
sglang serve \
  --model-path /models/MiniMax-H3 \
  --model-variant fl2va \
  --num-gpus 1 \
  --performance-mode memory \
  --layerwise-offload-components dit,text_encoder,vae \
  --dit-offload-prefetch-size 1 \
  --dit-layerwise-resident-layers 4 \
  --layerwise-resident-layers video_vae=36 \
  --enable-torch-compile false \
  --attention-backend torch_sdpa
```

A standalone allocator reproducer for the underlying race is:

```python
import gc
import torch

current = torch.cuda.current_stream()
courier = torch.cuda.Stream()
elements = 32 * 1024 * 1024

with torch.cuda.stream(courier):
    weight = torch.full((elements,), 1.0, device="cuda")
    ready = courier.record_event()

current.wait_event(ready)
torch.cuda._sleep(200_000_000)
output = weight.clone()
original_ptr = weight.data_ptr()
del weight
gc.collect()

with torch.cuda.stream(courier):
    replacement = torch.empty(elements, device="cuda")
    print("storage reused:", replacement.data_ptr() == original_ptr)
    replacement.fill_(7.0)

torch.cuda.synchronize()
print("correct:", bool(torch.all(output == 1.0)))
```

On the affected L20, the allocator reused the storage and the output was corrupted. Adding `weight.record_stream(current)` after `wait_event()` prevented reuse and restored correctness.

Full-model A/B validation used the same model revision, prompt, seed, 20 steps, 480p target, and `torch_sdpa` backend:

- disabling the courier produced a correct 15-second MP4;
- enabling the courier without lifetime registration produced colored noise;
- enabling the courier with `gpu_tensor.record_stream(compute_stream)` produced a correct MP4 whose SHA-256 was byte-identical to the courier-disabled baseline: `d66baa8ab742b1be5068cb14d8d35c01bc81f2dc43cc4ba8e9ea2d2e895a7701`.

## Expected behavior

Mapped weights shipped by the courier remain live until all compute-stream work using them has completed. Enabling the courier should not change model output.

## Environment

Relevant `sglang.check_env` / deployment probe values:

```text
OS: Linux
Python: 3.11
SGLang: 0.0.0.dev0
Reproduced SGLang commit: fe3d4b9bbbff744d056dda122ea9157c2932a2bd
Latest main inspected: dae126d510b786a2424684c17ca4e5e70d74e8f7
PyTorch: 2.13.0+cu130
CUDA runtime/compatibility: 13.0
NVIDIA driver: 535
GPU: NVIDIA L20, 46068 MiB
Ray: 2.49.2
Model: MiniMaxAI/MiniMax-H3
Model revision: 42ed227ee7df40d41602854ae760620d6eb651fe
Attention backend: torch_sdpa
```

I will submit a small fix and regression test.
